### PR TITLE
Update Infineon CYW30739 TE8 fixes.

### PR DESCRIPTION
**Pull request by hsusid**
#### Problem
* The platform CYW30739 needs updates to support TE8.

#### Change overview
* DoFactoryReset would clear the key-value storage.
* Enable OTA requestor for CYW30739 apps.
* Refine the flash layout for the OTA functions.
* Disable loggings for lighting and lock app.
* Reuse Matter heap APIs for CYW30739 OpenThread implementation.
* Enable MBEDTLS_CCM_ALT, MBEDTLS_AES_ALT, and MBEDTLS_SHA256_ALT.
* Group keys into factory keys and config keys.
* Fix CYW30739Config write methods.
* Correct the storage entry count for additional keys.
* Add and process OTA image header for CYW30739.
* Implement network-interfaces and reboot-count.
* Fix currentLevel value mismatch.
* Implement PICS_MANUAL_OCCUPANCY_CHANGE.
* Instantiate identify object for lighting app.
* Implement IsFirstImageRun and ConfirmCurrentImage methods of OTAImageProcessorImpl.
* Compress large "Data" entries in DS.
* Change CHIP_CONFIG_MAX_FABRICS to 5 from 4.
* Add a debug arg for CYW30739 SDK.
* Implement SupportedCalendarTypes and SupportedLocales.
* Implement Door lock cluster.

#### Testing
* TE8 test cases.

### CODE REVIEW

File: examples/lock-app/cyw30739/include/AppEvent.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+struct AppEvent;
```
Comment (andy31415) [inline examples/lock-app/cyw30739/include/AppEvent.h:24]: This declaration does not seem to be useful: AppEvent is declared just 2 lines below.

Comment (andy31415) [inline examples/lock-app/cyw30739/include/AppEvent.h:24]: This declaration does not seem to be useful: AppEvent is declared just 2 lines below.

File: examples/lock-app/cyw30739/src/ZclCallbacks.cpp [hunk 2]
```diff
@@ -61,3 +62,94 @@ void MatterPostAttributeChangeCallback(const app::ConcreteAttributePath & attrib
...
     printf("ERROR clusterId: 0x%04lx, unknown attribute ID: 0x%04lx\n", attributePath.mClusterId, attributePath.mAttributeId);
 }
+
+/** @brief DoorLock Cluster Init
+ *
+ * This function is called when a specific cluster is initialized. It gives the
+ * application an opportunity to take care of cluster initialization procedures.
+ * It is called exactly once for each endpoint where cluster is present.
+ *
+ * @param endpoint   Ver.: always
```
Comment (andy31415) [inline examples/lock-app/cyw30739/src/ZclCallbacks.cpp:72]: what does `Ver.: always` mean?

I can read the code as `this takes a single parameter representing the endpoint id` ... the comment confuses me more than it helps.

Comment (andy31415) [inline examples/lock-app/cyw30739/src/ZclCallbacks.cpp:72]: what does `Ver.: always` mean?

I can read the code as `this takes a single parameter representing the endpoint id` ... the comment confuses me more than it helps.

File: examples/platform/cyw30739/OTAConfig.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+class OTAConfig
+{
+public:
+    OTAConfig(){};
```
Comment (andy31415) [inline examples/platform/cyw30739/OTAConfig.h:24]: `;` at the end not required.



Why do we have this class containing only a static method? Why not use a namespace instead?

File: src/platform/CYW30739/DeviceInfoProviderImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+#include <lib/support/CodeUtils.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <platform/CYW30739/DeviceInfoProviderImpl.h>
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+namespace chip {
+namespace DeviceLayer {
```
Comment (andy31415) [inline src/platform/CYW30739/DeviceInfoProviderImpl.cpp:29]: This is not the right place to implement these. This is subject to changes, see #18687

Comment (andy31415) [inline src/platform/CYW30739/DeviceInfoProviderImpl.cpp:29]: This is not the right place to implement these. This is subject to changes, see #18687

File: src/platform/CYW30739/DiagnosticDataProviderImpl.cpp [hunk 5]
```diff
@@ -65,5 +69,49 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapHighWatermark(uint64_t & cu
...
+    {
+        // If the value overflows, return UINT16 max value to provide best-effort number.
+        rebootCount = static_cast<uint16_t>(count <= UINT16_MAX ? count : UINT16_MAX);
+    }
+    return err;
+}
+
+CHIP_ERROR DiagnosticDataProviderImpl::GetNetworkInterfaces(NetworkInterface ** netifpp)
+{
+    NetworkInterface * ifp = new NetworkInterface();
```
Comment (andy31415) [inline src/platform/CYW30739/DiagnosticDataProviderImpl.cpp:86]: why not use Platform::New and Delete?

Comment (andy31415) [inline src/platform/CYW30739/DiagnosticDataProviderImpl.cpp:86]: why not use Platform::New and Delete?

File: src/platform/CYW30739/KeyValueStoreManagerImpl.cpp [hunk 6]
```diff
@@ -128,6 +128,20 @@ CHIP_ERROR KeyValueStoreManagerImpl::_Delete(const char * key)
...
 
+CHIP_ERROR KeyValueStoreManagerImpl::EraseAll(void)
+{
+    wiced_result_t result;
+    wiced_hal_delete_nvram(PLATFORM_NVRAM_ID_ENTRY_INFO_STRING, &result);
+    if (result == WICED_SUCCESS)
+    {
+        return CHIP_NO_ERROR;
+    }
+    else
```
Comment (andy31415) [inline src/platform/CYW30739/KeyValueStoreManagerImpl.cpp:139]: nit: prefer no else after return.

maybe even



```cpp

return (result == WICED_SUCCESS) ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_FAILED

```



you may want to actually print out the result in case of failure though as you lose the actual result error code here and return a generic `failed` status which is hard to debug.

Then again, logs seem disabled for this platform :/

Comment (andy31415) [inline src/platform/CYW30739/KeyValueStoreManagerImpl.cpp:139]: nit: prefer no else after return.

maybe even



```cpp

return (result == WICED_SUCCESS) ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_FAILED

```



you may want to actually print out the result in case of failure though as you lose the actual result error code here and return a generic `failed` status which is hard to debug.

Then again, logs seem disabled for this platform :/

File: third_party/cyw30739_sdk/compress_cgs_data.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    cgs = parse_cgs(option.input)
+    compress_cgs(option, cgs)
+    export_cgs(option.output, cgs)
+
+
+def parse_cgs(file_name):
+    cgs = []
+
+    with open(file_name, mode="r") as file:
+        while True:
```
Comment (andy31415) [inline third_party/cyw30739_sdk/compress_cgs_data.py:35]: why not `for line in file.readlines()` ?

Comment (andy31415) [inline third_party/cyw30739_sdk/compress_cgs_data.py:35]: why not `for line in file.readlines()` ?

File: third_party/cyw30739_sdk/compress_cgs_data.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                item_type = "definition"
+
+            cgs.append(Item(line, item_type, nested_data))
+
+    return cgs
+
+
+def parse_entry(file: io.TextIOBase, line):
+
+    data_re = re.compile(r'^\s*ENTRY\s+"([^"]+)"(?:\s*=\s*"([^"]+)")?$')
```
Comment (andy31415) [inline third_party/cyw30739_sdk/compress_cgs_data.py:59]: you may want to consider writing some parser like lark based parsers (IDL parsers use this). this is reasonably hard to follow: at least a comment explaining what this parses and how would be helpful: example of how a line looks like and what parsing does.

Comment (andy31415) [inline third_party/cyw30739_sdk/compress_cgs_data.py:59]: you may want to consider writing some parser like lark based parsers (IDL parsers use this). this is reasonably hard to follow: at least a comment explaining what this parses and how would be helpful: example of how a line looks like and what parsing does.

File: third_party/cyw30739_sdk/merge_hex.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
```
Comment (andy31415) [inline third_party/cyw30739_sdk/merge_hex.py:1]: license blurb

also comment on what it does.

File: examples/lock-app/cyw30739/include/AppEvent.h (hunk 1 above)
Comment (hsusid) [reply examples/lock-app/cyw30739/include/AppEvent.h]: removed.

File: examples/lock-app/cyw30739/src/ZclCallbacks.cpp (hunk 2 above)
Comment (hsusid) [reply examples/lock-app/cyw30739/src/ZclCallbacks.cpp]: removed.

File: examples/platform/cyw30739/OTAConfig.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsusid) [reply examples/platform/cyw30739/OTAConfig.h]: has turned OTAConfig into a namespace.

File: src/platform/CYW30739/DeviceInfoProviderImpl.cpp (hunk 4 above)
Comment (hsusid) [reply src/platform/CYW30739/DeviceInfoProviderImpl.cpp]: reverted.

File: src/platform/CYW30739/DiagnosticDataProviderImpl.cpp (hunk 5 above)
Comment (hsusid) [reply src/platform/CYW30739/DiagnosticDataProviderImpl.cpp]: done.

File: src/platform/CYW30739/KeyValueStoreManagerImpl.cpp (hunk 6 above)
Comment (hsusid) [reply src/platform/CYW30739/KeyValueStoreManagerImpl.cpp]: fixed.

File: third_party/cyw30739_sdk/compress_cgs_data.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (hsusid) [reply third_party/cyw30739_sdk/compress_cgs_data.py:1]: done.

File: third_party/cyw30739_sdk/compress_cgs_data.py (hunk 7 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/compress_cgs_data.py]: fixed.

File: third_party/cyw30739_sdk/compress_cgs_data.py (hunk 8 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/compress_cgs_data.py]: I have reformated codes to improve readability.
I wish I had known lark before I implemented this script.

File: third_party/cyw30739_sdk/merge_hex.py (hunk 9 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/merge_hex.py:1]: done.

File: third_party/cyw30739_sdk/gen_ota_image.py [hunk 11]
```diff
@@ -84,5 +99,89 @@ def pad_data(data, aligned_size):
...
+        check=True,
+        capture_output=True,
+        text=True,
+    ).stdout.splitlines():
+        if line.startswith("Header Size:"):
+            header_size = int(line.split(":")[1])
+            break
+
+    if header_size % 4 == 0:
+        return 0
```
Comment (andy31415) [inline third_party/cyw30739_sdk/gen_ota_image.py:122]: is there any value in returning anything here if this method always returns 0?

File: third_party/cyw30739_sdk/merge_hex.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+#
+# Copyright 2016-2022, Cypress Semiconductor Corporation (an Infineon company) or
```
Comment (andy31415) [inline third_party/cyw30739_sdk/merge_hex.py:3]: if this is in CHIP code, should contain CHIP license attribution.

File: third_party/cyw30739_sdk/compress_cgs_data.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+#
+# Copyright 2016-2022, Cypress Semiconductor Corporation (an Infineon company) or
```
Comment (andy31415) [inline third_party/cyw30739_sdk/compress_cgs_data.py:3]: Code in CHIP should be attributed to CHIP authors, not Cypress/Infineon. I believe it may be dual licensed however it cannot be just infineon.

File: third_party/cyw30739_sdk/compress_cgs_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply third_party/cyw30739_sdk/compress_cgs_data.py:1]: CHIP license blurb

[the remark this answers is not in the document]
Comment (hsusid) [reply third_party/cyw30739_sdk/compress_cgs_data.py:1]: done.

File: third_party/cyw30739_sdk/gen_ota_image.py (hunk 11 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/gen_ota_image.py]: fixed.

File: third_party/cyw30739_sdk/merge_hex.py (hunk 12 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/merge_hex.py]: done.

File: third_party/cyw30739_sdk/compress_cgs_data.py (hunk 13 above)
Comment (hsusid) [reply third_party/cyw30739_sdk/compress_cgs_data.py]: done.

**Review by andy31415 — APPROVED** [review]

**Comment by hsusid** [comment]
rebase

**Comment by hsusid** [comment]
rebase again to see if CI would be fixed.

File: src/platform/CYW30739/DiagnosticDataProviderImpl.cpp [hunk 14]
```diff
@@ -65,5 +69,49 @@ CHIP_ERROR DiagnosticDataProviderImpl::GetCurrentHeapHighWatermark(uint64_t & cu
...
+    ifp->isOperational             = true;
+    ifp->offPremiseServicesReachableIPv4.SetNull();
+    ifp->offPremiseServicesReachableIPv6.SetNull();
+    ifp->type = app::Clusters::GeneralDiagnostics::InterfaceType::EMBER_ZCL_INTERFACE_TYPE_THREAD;
+#else
+    /* TODO */
+#endif
+    uint8_t macBuffer[ConfigurationManager::kPrimaryMACAddressLength];
+    ConfigurationMgr().GetPrimary802154MACAddress(macBuffer);
+    ifp->hardwareAddress = ByteSpan(macBuffer, ConfigurationManager::kPrimaryMACAddressLength);
```
Comment (bzbarsky-apple) [inline src/platform/CYW30739/DiagnosticDataProviderImpl.cpp:100]: This is not really OK.  `ifp->hardwareAddress` now points to the stack buffer, and once this function returns that will be a dangling pointer.  Please do a followup to fix.

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/lighting-app/cyw30739/src/ZclCallbacks.cpp [hunk 15]
```diff
@@ -63,6 +63,10 @@ void MatterPostAttributeChangeCallback(const app::ConcreteAttributePath & attrib
             return;
         }
         break;
+    case Identify::Id:
+        ChipLogProgress(Zcl, "Identify attribute ID: " ChipLogFormatMEI " Type: %u Value: %u, length %u",
+                        ChipLogValueMEI(attributePath.mClusterId), type, *value, size);
```
[the remark this answers is not in the document]
Comment (hsusid) [reply examples/lighting-app/cyw30739/src/ZclCallbacks.cpp:68]: https://github.com/project-chip/connectedhomeip/pull/19568
